### PR TITLE
Add core::Result and tl::expected

### PR DESCRIPTION
**Pull request by dmortondev**
* Incorporate backwards-compatible std::expected implementation.
* Add Posit-specific wrapper types that bind the error type of expected to core::Error.

### Intent

> Describe briefly what problem this pull request resolves, or what new feature it introduces. Include screenshots of any new or altered UI. Link to any Github issues that are related. 

Upstream `core::Result` from Launcher repo so it's available to a wider group of developers. No user-facing impact.

### Approach

> Describe the approach taken and the tradeoffs involved if non-obvious; add an overview of the solution if it's complicated.

I did add `expected.hpp` directly to the repo instead of pulling it in as a git submodule. It's a single file and new versions are few and far between.

### Automated Tests

> Indicate whether this change includes unit tests or integration tests, or link a work item or pull request to add those tests to another repo. If the change cannot or will not be covered by a test, indicate why.

n/a: no functional impact.

### QA Notes

> Add additional information for QA on how to validate the change, paying special attention to the level of risk, adjacent areas that could be affected by the change, and any important contextual information not present in the linked issues. 

n/a: no functional impact.

n/a: no user impact.

### Checklist

- [ ] If this PR adds a new feature, or fixes a bug in a previously released version, it includes an entry in `NEWS.md` 
- [ ] If this PR adds or changes UI, the updated UI meets [accessibility standards](https://github.com/rstudio/rstudio/wiki/Accessibility)
- [ ] A reviewer is assigned to this PR (if unsure who to assign, check Area Owners list)
- [ ] This PR passes all local unit tests

### CODE REVIEW

File: src/cpp/core/include/core/Result.hpp [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+ *
+ */
+
+#ifndef CORE_RESULT_HPP
+#define CORE_RESULT_HPP
+
+#include <expected/expected.hpp>
+#include <shared_core/Error.hpp>
+
+namespace rstudio::core {
```
Comment (cm421) [inline src/cpp/core/include/core/Result.hpp:22]:
```suggestion
namespace rstudio {
namespace core {
```

This syntax is a C++17 feature :(

File: src/cpp/core/include/core/Result.hpp [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+// {
+//    if (x < 0) {return core::Unexpected(core::Error("x must be non-negative"));}
+//    return std::sqrt(x);
+// }
+template <typename T> using Result = tl::expected<T, Error>;
+
+// Type alias for the unexpected (error) type of core::Result.
+using Unexpected = tl::unexpected<Error>;
+
+} // namespace rstudio::core
```
Comment (cm421) [inline src/cpp/core/include/core/Result.hpp:41]:
```suggestion
} // core
} // rstudio
```

**Review by cm421 — APPROVED** [review]
I just confirmed with Maria that one last thing we'll need to do is add an entry for this `expected.hpp` code into the `NOTICE` file at the root directory of our repo. 

Add `tl::expected` (or whatever you'd like it to be called) to the list at the top of the file, and then a copy of the CC0 Public Domain Dedication license in that same file and then we should be good to go!

**Review by zachhannum — APPROVED** [review]
I think now is a good time to get this in, towards the beginning of a release cycle!

**Pull request merged**